File: vscode/extension/src/lsp/custom.ts
export interface AllModelsMethod {
  method: 'sqlmesh/all_models'
  request: AllModelsRequest
  response: AllModelsResponse
}

export interface RenderModelMethod {
  method: 'sqlmesh/render_model'
  request: RenderModelRequest
  response: RenderModelResponse
}

interface RenderModelRequest {
  textDocumentUri: string
}

interface RenderModelResponse extends BaseResponse {
  models: RenderModelEntry[]
}

export interface RenderModelEntry {
  name: string
  fqn: string
  description: string | null | undefined
  rendered_query: string
}

export type CustomLSPMethods =
  | AllModelsMethod
  | AbstractAPICall
  | RenderModelMethod
  | AllModelsForRenderMethod
  | SupportedMethodsMethod
  | FormatProjectMethod
  | ListWorkspaceTests
  | ListDocumentTests
  | RunTest
  | GetEnvironmentsMethod
  | GetTableDiffModelsMethod

interface AllModelsRequest {
  textDocument: {
    uri: string
  }
}

interface AllModelsResponse extends BaseResponse {
  models: string[]
  keywords: string[]
}

export interface AbstractAPICallRequest {
  url: string
  method: string
  params: Record<string, any>
  body: Record<string, any>
}

export interface AbstractAPICall {
  method: 'sqlmesh/api'
  request: AbstractAPICallRequest
  response: AbstractAPICallResponse
}

type AbstractAPICallResponse = object & BaseResponse

export interface AllModelsForRenderMethod {
  method: 'sqlmesh/all_models_for_render'
  request: AllModelsForRenderRequest
  response: AllModelsForRenderResponse
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
interface AllModelsForRenderRequest {}

interface AllModelsForRenderResponse extends BaseResponse {
  models: ModelForRendering[]
}

export interface ModelForRendering {
  name: string
  fqn: string
  description: string | null | undefined
  uri: string
}

export interface SupportedMethodsMethod {
  method: 'sqlmesh/supported_methods'
  request: SupportedMethodsRequest
  response: SupportedMethodsResponse
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
interface SupportedMethodsRequest {}

interface SupportedMethodsResponse extends BaseResponse {
  methods: CustomMethod[]
}

interface CustomMethod {
  name: string
}

export interface FormatProjectMethod {
  method: 'sqlmesh/format_project'
  request: FormatProjectRequest
  response: FormatProjectResponse
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
interface FormatProjectRequest {}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
interface FormatProjectResponse extends BaseResponse {}

interface BaseResponse {
  response_error?: string
}

export interface ListWorkspaceTests {
  method: 'sqlmesh/list_workspace_tests'
  request: ListWorkspaceTestsRequest
  response: ListWorkspaceTestsResponse
}

type ListWorkspaceTestsRequest = object

interface Position {
  line: number
  character: number
}

interface Range {
  start: Position
  end: Position
}

interface TestEntry {
  name: string
  uri: string
  range: Range
}

interface ListWorkspaceTestsResponse extends BaseResponse {
  tests: TestEntry[]
}

export interface ListDocumentTests {
  method: 'sqlmesh/list_document_tests'
  request: ListDocumentTestsRequest
  response: ListDocumentTestsResponse
}

export interface DocumentIdentifier {
  uri: string
}

export interface ListDocumentTestsRequest {
  textDocument: DocumentIdentifier
}

export interface ListDocumentTestsResponse extends BaseResponse {
  tests: TestEntry[]
}

export interface RunTest {
  method: 'sqlmesh/run_test'
  request: RunTestRequest
  response: RunTestResponse
}

export interface RunTestRequest {
  textDocument: DocumentIdentifier
  testName: string
}

export interface RunTestResponse extends BaseResponse {
  success: boolean
  error_message?: string
}

export interface GetEnvironmentsMethod {
  method: 'sqlmesh/get_environments'
  request: GetEnvironmentsRequest
  response: GetEnvironmentsResponse
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
interface GetEnvironmentsRequest {}

interface GetEnvironmentsResponse extends BaseResponse {
  environments: Record<string, EnvironmentInfo>
  pinned_environments: string[]
  default_target_environment: string
}

interface EnvironmentInfo {
  name: string
  snapshots: string[]
  start_at: string
  plan_id: string
}

export interface GetTableDiffModelsMethod {
  method: 'sqlmesh/get_models'
  request: GetModelsRequest
  response: GetModelsResponse
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
interface GetModelsRequest {}

interface GetModelsResponse extends BaseResponse {
  models: ModelInfo[]
}

interface ModelInfo {
  name: string
  fqn: string
  description: string | null | undefined
}

File: sqlmesh/lsp/custom.py
from lsprotocol import types
import typing as t

from sqlmesh.core.linter.rule import Range
from sqlmesh.utils.pydantic import PydanticModel


class CustomMethodRequestBaseClass(PydanticModel):
    pass


class CustomMethodResponseBaseClass(PydanticModel):
    # Prefixing, so guaranteed not to collide
    response_error: t.Optional[str] = None


ALL_MODELS_FEATURE = "sqlmesh/all_models"


class AllModelsRequest(CustomMethodRequestBaseClass):
    """
    Request to get all the models that are in the current project.
    """

    textDocument: types.TextDocumentIdentifier


class MacroCompletion(PydanticModel):
    """Information about a macro for autocompletion."""

    name: str
    description: t.Optional[str] = None


class ModelCompletion(PydanticModel):
    """Information about a model for autocompletion."""

    name: str
    description: t.Optional[str] = None


class AllModelsResponse(CustomMethodResponseBaseClass):
    """Response to get all models that are in the current project."""

    #: Deprecated: use ``model_completions`` instead
    models: t.List[str]
    model_completions: t.List[ModelCompletion]
    keywords: t.List[str]
    macros: t.List[MacroCompletion]


RENDER_MODEL_FEATURE = "sqlmesh/render_model"


class RenderModelRequest(CustomMethodRequestBaseClass):
    textDocumentUri: str


class RenderModelEntry(PydanticModel):
    """
    An entry in the rendered model.
    """

    name: str
    fqn: str
    description: t.Optional[str] = None
    rendered_query: str


class RenderModelResponse(CustomMethodResponseBaseClass):
    """
    Response to render a model.
    """

    models: t.List[RenderModelEntry]


ALL_MODELS_FOR_RENDER_FEATURE = "sqlmesh/all_models_for_render"


class ModelForRendering(PydanticModel):
    """
    A model that is available for rendering.
    """

    name: str
    fqn: str
    description: t.Optional[str] = None
    uri: str


class AllModelsForRenderRequest(CustomMethodRequestBaseClass):
    pass


class AllModelsForRenderResponse(CustomMethodResponseBaseClass):
    """
    Response to get all the models that are in the current project for rendering purposes.
    """

    models: t.List[ModelForRendering]


SUPPORTED_METHODS_FEATURE = "sqlmesh/supported_methods"


class SupportedMethodsRequest(PydanticModel):
    """
    Request to get all supported custom LSP methods.
    """

    pass


class CustomMethod(PydanticModel):
    """
    Information about a custom LSP method.
    """

    name: str


class SupportedMethodsResponse(CustomMethodResponseBaseClass):
    """
    Response containing all supported custom LSP methods.
    """

    methods: t.List[CustomMethod]


FORMAT_PROJECT_FEATURE = "sqlmesh/format_project"


class FormatProjectRequest(CustomMethodRequestBaseClass):
    """
    Request to format all models in the current project.
    """

    pass


class FormatProjectResponse(CustomMethodResponseBaseClass):
    """
    Response to format project request.
    """

    pass


LIST_WORKSPACE_TESTS_FEATURE = "sqlmesh/list_workspace_tests"


class ListWorkspaceTestsRequest(CustomMethodRequestBaseClass):
    """
    Request to list all tests in the current project.
    """

    pass


GET_ENVIRONMENTS_FEATURE = "sqlmesh/get_environments"


class GetEnvironmentsRequest(CustomMethodRequestBaseClass):
    """
    Request to get all environments in the current project.
    """

    pass


class TestEntry(PydanticModel):
    """
    An entry representing a test in the workspace.
    """

    name: str
    uri: str
    range: Range


class ListWorkspaceTestsResponse(CustomMethodResponseBaseClass):
    tests: t.List[TestEntry]


LIST_DOCUMENT_TESTS_FEATURE = "sqlmesh/list_document_tests"


class ListDocumentTestsRequest(CustomMethodRequestBaseClass):
    textDocument: types.TextDocumentIdentifier


class ListDocumentTestsResponse(CustomMethodResponseBaseClass):
    tests: t.List[TestEntry]


RUN_TEST_FEATURE = "sqlmesh/run_test"


class RunTestRequest(CustomMethodRequestBaseClass):
    textDocument: types.TextDocumentIdentifier
    testName: str


class RunTestResponse(CustomMethodResponseBaseClass):
    success: bool
    error_message: t.Optional[str] = None


class EnvironmentInfo(PydanticModel):
    """
    Information about an environment.
    """

    name: str
    snapshots: t.List[str]
    start_at: str
    plan_id: str


class GetEnvironmentsResponse(CustomMethodResponseBaseClass):
    """
    Response containing all environments in the current project.
    """

    environments: t.Dict[str, EnvironmentInfo]
    pinned_environments: t.Set[str]
    default_target_environment: str


GET_MODELS_FEATURE = "sqlmesh/get_models"


class GetModelsRequest(CustomMethodRequestBaseClass):
    """
    Request to get all models available for table diff.
    """

    pass


class ModelInfo(PydanticModel):
    """
    Information about a model for table diff.
    """

    name: str
    fqn: str
    description: t.Optional[str] = None


class GetModelsResponse(CustomMethodResponseBaseClass):
    """
    Response containing all models available for table diff.
    """

    models: t.List[ModelInfo]
